mlrun/db/sqldb.py

```diff
@@ -0,0 +1,289 @@
+# Copyright 2019 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pickle
+from datetime import datetime, timedelta
+
+from sqlalchemy import BLOB, TIMESTAMP, Column, Integer, String, create_engine
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+
+from ..lists import ArtifactList, FunctionList, RunList
+from ..utils import get_in, update_in
+from .base import RunDBError, RunDBInterface
+
+Base = declarative_base()
+
+
+class HasStruct:
+    @property
+    def struct(self):
+        return pickle.loads(self.body)
+
+    @struct.setter
+    def struct(self, value):
+        self.body = pickle.dumps(value)
+
+
+# TODO: Store labels in db (one to many)
+class Artifact(Base, HasStruct):
+    __tablename__ = 'artifacts'
+
+    id = Column(Integer, primary_key=True)
+    key = Column(String)
+    project = Column(String)
+    tag = Column(String)
+    uid = Column(String)
+    updated = Column(TIMESTAMP)
+    body = Column(BLOB)
+
+
+class Function(Base, HasStruct):
+    __tablename__ = 'functions'
+
+    id = Column(Integer, primary_key=True)
+    name = Column(String)
+    project = Column(String)
+    tag = Column(String)
+    body = Column(BLOB)
+
+
+class Log(Base):
+    __tablename__ = 'logs'
+
+    id = Column(Integer, primary_key=True)
+    uid = Column(String)
+    project = Column(String)
+    body = Column(BLOB)
+
+
+class Run(Base, HasStruct):
+    __tablename__ = 'runs'
+    id = Column(Integer, primary_key=True)
+    uid = Column(String)
+    project = Column(String)
+    body = Column(BLOB)
+
+
+class SQLDB(RunDBInterface):
+    def __init__(self, dsn):
+        self.dsn = dsn
+        self.session = None
+
+    def connect(self, secrets=None):
+        engine = create_engine(self.dsn)
+        Base.metadata.create_all(engine)
+        cls = sessionmaker(bind=engine)
+        # TODO: One session per call?
+        self.session = cls()
+
+    def store_log(self, uid, project='', body=None, append=True):
```

Comment:
@tebeka for logs we should keep it on a file path (since logs can be big and db not efficient, no queries..

Reply:
> we should keep it on a file path 

Where? If we're running on several machines this means we'll have the logs in different places and then use can store a log and later fail to get it since it'll request from another machine.
Maybe have it in another issues?

> logs can be big 
How big? What' the median log size? 
It's text and we can store them compressed in the database.

Reply:
@tebeka just make it configurable via env, we can always mount on our clustered FS if we need shared, no need to overload the DB w logs

Reply:
OK, opened #75